### ISSUE TITLE
Allowing different types of matrices in MarkovChains

### ISSUE THREAD

**Issue by czgdp1807** [sympy/sympy#20448]
This idea was earlier was suggested by @Maelstrom6 but due to preserve simplicity of API and it's implementation I wanted to avoid that. However, we can satisfy both the requirements of simplicity and flexibility by adding a conversion layer for different types of matrices. Basically the following should happen,

1. Any Concrete Matrix -> Conversion Layer -> `sympy.Matrix` -> Existing query handler for MarkovChains.
2. Any Symbolic Matrix -> Conversion Layer -> `sympy.MatrixSymbol` -> Existing query handler for MarkovChains

This way the algorithm will be kept untouched and the API will become more flexible. I don't know if anybody has suggested this earlier or not, opening this issue so that we don't forget to work on this.

**Comment by oscarbenjamin** [comment]
Can you demonstrate how the API works that takes matrices but does not work for matrix expressions?

**Comment by czgdp1807** [comment]
I forgot about matrix expressions. In fact that can also be handeled by the conversion layer.

**Comment by Maelstrom6** [comment]
Most functions and APIs need an explicit matrix, since things like `P`, `E` and `communication_classes` all use explicit indexing.

**Comment by oscarbenjamin** [comment]
It is possible to index a matrix expression if the size is not symbolic. However in general would it not be possible for some of these operations to be unevaluated in terms of a matrix expression?

**Comment by Maelstrom6** [comment]
> However in general would it not be possible for some of these operations to be unevaluated in terms of a matrix expression?

I don't think so. `P(Eq(X[0], 0))` needs the 0th row of the matrix. It could probably be adapted for `FunctionMatrix` but not anything else. `communication_classes` uses `strongly_connected_components` which cannot take a symbolic number of arguments.

**Comment by oscarbenjamin** [comment]
I think that `P` can be unevaluated.

We can extract a row of a matrix with
```
In [5]: Matrix([1, 0]).T * Matrix([[1, 2], [3, 4]])                                                                               
Out[5]: [1  2]
```
There's probably another way to do that symbolically with other classes.

We also have `BlockMatrix` for which the connected components can be identified at the block level potentially.

**Comment by czgdp1807** [comment]
I think unevaluated results can act as a fall back if conversion layer cannot convert the given matrix to either `Matrix` or `MatrixSymbol` or any index-able matrix.

**Comment by oscarbenjamin** [comment]
Most matrix expressions can be indexed if the shape is explicit.

**Comment by czgdp1807** [comment]
That's great. So, we need to implement the conversion layer and connect it with the current algorithm.

**Comment by ARUN1321** [comment]
Can i get an idea in MarkovChains please

**Comment by ARUN1321** [comment]
And i have idea in creating different types matrix like 
Square Matrix.
Symmetric Matrix.
Triangular Matrix.
Diagonal Matrix.
Identity Matrix.
Orthogonal Matrix.

This is what should I do
